src/node_options.cc

```diff
@@ -0,0 +1,335 @@
+#if !defined(NODE_HAVE_SMALL_ICU)
+         "\t                          (will extend linked-in data)\n"
+#endif
+#endif
+         "\n"
+         "Documentation can be found at https://iojs.org/\n");
+}
+
+void NodeOptions::ParseArgs(int* argc,
+                      const char** argv,
+                      int* exec_argc,
+                      const char*** exec_argv,
+                      int* v8_argc,
+                      const char*** v8_argv) {
+  const unsigned int nargs = static_cast<unsigned int>(*argc);
+  const char** new_exec_argv = new const char*[nargs];
+  const char** new_v8_argv = new const char*[nargs];
+  const char** new_argv = new const char*[nargs];
+  const char** local_preload_modules = new const char*[nargs];
+
+  char** argv_ = const_cast<char**>(argv);
+  for (unsigned int i = 0; i < nargs; ++i) {
+    new_exec_argv[i] = nullptr;
+    new_v8_argv[i] = nullptr;
+    new_argv[i] = nullptr;
+    local_preload_modules[i] = nullptr;
+  }
+
+  const char* port = nullptr;
+
+  // exec_argv starts with the first option, the other two start with argv[0].
+  unsigned int new_exec_argc = 0;
+  unsigned int new_v8_argc = 1;
+  unsigned int new_argc = 1;
+  new_v8_argv[0] = argv[0];
+  new_argv[0] = argv[0];
+  int rc = 0;
+  unsigned int index = 1;
+  bool is_eval = false;
+  bool is_print = false;
+  const char* optstring = ":hve:p:ir:d::b::x:";
+  while ((rc = getopt_long(*argc, argv_, optstring, longopts, NULL)) != -1 &&
+          argv[index][0] == '-') {
+    unsigned int args_consumed = 1;
+    const char* const arg = argv[index];
+    switch (rc) {
+      case 'h':
+        PrintHelp();
+        exit(0);
+        break;
+      case 'v':
+        printf("%s\n", NODE_VERSION);
+        exit(0);
+        break;
+      case 'e':
+      case 'p':
+      {
+        if (!is_eval)
+          is_eval = (rc == 'e');
+
+        if (!is_print)
+          is_print = (rc == 'p');
+        const char* name = is_eval ? "eval" : "print";
+        NodeOptions::print_eval = NodeOptions::print_eval || is_print;
+        if (is_eval == true) {
+          NodeOptions::eval_string = argv[index + 1];
+          args_consumed += 1;
+          if (NodeOptions::eval_string == nullptr) {
+            fprintf(stderr, "%s: %s requires an argument\n", argv[0], name);
+            exit(9);
+          }
+        } else {
+          if ((index + 1 < nargs) &&
+              argv[index + 1] != nullptr &&
+              argv[index + 1][0] != '-') {
+            NodeOptions::eval_string = argv[index + 1];
+            args_consumed += 1;
+            if (strncmp(NodeOptions::eval_string, "\\-", 2) == 0) {
+              // Starts with "\\-": escaped expression, drop the backslash.
+              NodeOptions::eval_string += 1;
+            }
+          } else if ((index + 2 < nargs) &&
+                      argv[index + 2] != nullptr &&
+                      argv[index + 2][0] != '-') {
```

Comment:
The current code does not do that. That was done for the case of `-p -e <arg>`. Really gross though